src/thread_pool_async.ml

```diff
@@ -2,83 +2,113 @@ open Base
 open Async_kernel
 open Async_unix
 
-type 'state thread = In_thread.Helper_thread.t * 'state
+type 'state worker = {
+  thread : In_thread.Helper_thread.t;
+  state : 'state;
+}
+
+exception Invalid_thread_count of int
+
+exception Pool_already_destroyed
+
+exception State_manipulation_error of exn
 
 type 'state t = {
-  pool : 'state thread Pipe.Reader.t * 'state thread Pipe.Writer.t;
+  reader : 'state worker Pipe.Reader.t;
+  writer : 'state worker Pipe.Writer.t;
   create : unit -> 'state;
   destroy : 'state -> unit;
-  threads : int;
+  thread_count : int;
 }
 
-let init ~name ~threads ~create ~destroy =
-  let create_thread ~name ~create =
-    let%bind thread = In_thread.Helper_thread.create ~name () in
-    let%bind state = In_thread.run ~thread create in
-    return (thread, state)
+let try_in_thread ~thread f = try_with (fun () -> In_thread.run ~thread f)
+
+let destroy_pool {reader; writer; thread_count; destroy; _} =
+  let%bind states_destroyed =
+    List.init thread_count ~f:ignore
+    |> Deferred.List.map ~how:`Parallel ~f:(fun () ->
+           match%bind Pipe.read reader with
+           | `Eof -> return @@ Ok ()
+           | `Ok {thread; state} -> try_in_thread ~thread (fun () -> destroy state))
+    |> Deferred.map ~f:Result.all_unit
```

Comment:
I believe it should be possible to use `Deferred.Result.all_unit` here.

Reply:
The signatures wouldn't match then.